### PR TITLE
vmware: do not create vm datastore when attach an existing volume on solidfire

### LINKED ISSUE #7148: [SolidFire] Cannot attach volume which is detached from VM
see title. The issue happens when use Solidfire as primary storage on vmware.

steps to reproduce the issue
1. create a volume and attach to vm
2. detach volume
3. attach volume again

Expected result: succeed
Actual result:

```
2023-01-27 14:40:19,380 ERROR [c.c.s.r.VmwareStorageProcessor] (DirectAgent-489:ctx-057df38b xxx.yyy.zzz, job-44722/job-44723, cmd: AttachCommand) (logid:736275cc) Failed to attach volume!
com.vmware.vim25.DuplicateNameFaultMsg: The name '-iqn.2010-01.com.solidfire:3p53.data-vol-test-01.62-0' already exists.
    at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
    at com.sun.xml.ws.fault.SOAPFaultBuilder.createException(SOAPFaultBuilder.java:117)
    at com.sun.xml.ws.client.sei.StubHandler.readResponse(StubHandler.java:223)
    at com.sun.xml.ws.db.DatabindingImpl.deserializeResponse(DatabindingImpl.java:176)
    at com.sun.xml.ws.db.DatabindingImpl.deserializeResponse(DatabindingImpl.java:263)
    at com.sun.xml.ws.client.sei.SyncMethodHandler.invoke(SyncMethodHandler.java:89)
    at com.sun.xml.ws.client.sei.SyncMethodHandler.invoke(SyncMethodHandler.java:62)
    at com.sun.xml.ws.client.sei.SEIStub.invoke(SEIStub.java:131)
    at com.sun.proxy.$Proxy698.createVmfsDatastore(Unknown Source)
    at com.cloud.hypervisor.vmware.mo.HostDatastoreSystemMO.createVmfsDatastore(HostDatastoreSystemMO.java:211)
    at com.cloud.storage.resource.VmwareStorageProcessor.getVmfsDatastore(VmwareStorageProcessor.java:2825)
    at com.cloud.storage.resource.VmwareStorageProcessor.prepareManagedDatastore(VmwareStorageProcessor.java:2666)
    at com.cloud.storage.resource.VmwareStorageProcessor.prepareManagedStorage(VmwareStorageProcessor.java:3450)
    at com.cloud.storage.resource.VmwareStorageProcessor.attachVolume(VmwareStorageProcessor.java:2034)
    at com.cloud.storage.resource.VmwareStorageProcessor.attachVolume(VmwareStorageProcessor.java:1997)
    at com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.execute(StorageSubsystemCommandHandlerBase.java:163)
    at com.cloud.storage.resource.StorageSubsystemCommandHandlerBase.handleStorageCommands(StorageSubsystemCommandHandlerBase.java:67)
    at com.cloud.hypervisor.vmware.resource.VmwareResource.executeRequest(VmwareResource.java:586)
    at com.cloud.agent.manager.DirectAgentAttache$Task.runInContext(DirectAgentAttache.java:315)
    at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:48)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:102)
    at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:52)
    at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:45)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:829)

```

As the datastore has been already created (before DetachVolume), cloudstack just needs to mount the datastore instead of creating a new datastore.

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
vmware, storage

~~~

##### CLOUDSTACK VERSION

~~~
4.17.2.0 (and probably previous versions)
~~~

##### CONFIGURATION

vmware
solidfire as primary storage

##### STEPS TO REPRODUCE

~~~

~~~

##### EXPECTED RESULTS

~~~

~~~

##### ACTUAL RESULTS

~~~

~~~

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR may fix #7148 , solidfire users are requested to test it.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test rocky8 vmware-70u3

**Comment by rohityadavcloud** [comment]
@blueorangutan package

File: plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java [hunk 1]
```diff
@@ -2822,7 +2822,10 @@ private ManagedObjectReference getVmfsDatastore(VmwareContext context, VmwareHyp
             throw new Exception("A relevant SCSI disk could not be located to use to create a datastore.");
         }
 
-        morDs = firstHostDatastoreSystemMO.createVmfsDatastore(datastoreName, hostScsiDisk);
+        morDs = firstHostDatastoreSystemMO.findDatastoreByName(datastoreName);
+        if (morDs == null) {
+            morDs = firstHostDatastoreSystemMO.createVmfsDatastore(datastoreName, hostScsiDisk);
```
Comment (rohityadavcloud) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java:2827]: LGTM, the code is trying to find a DS by name and when it's not found it's creating a vmfs datastore. Should this be generic and enough to merge/accept this change - @harikrishna-patnala @shwstppr @nvazquez  @DaanHoogland ?

**Review by DaanHoogland — APPROVED** [review]
cLGTM

File: plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java [hunk 2]
```diff
@@ -2822,7 +2822,10 @@ private ManagedObjectReference getVmfsDatastore(VmwareContext context, VmwareHyp
             throw new Exception("A relevant SCSI disk could not be located to use to create a datastore.");
         }
 
-        morDs = firstHostDatastoreSystemMO.createVmfsDatastore(datastoreName, hostScsiDisk);
+        morDs = firstHostDatastoreSystemMO.findDatastoreByName(datastoreName);
```
Comment (shwstppr) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java:2825]: Will it be better to use `HypervisorHostHelper.findDatastoreWithBackwardsCompatibility` instead? @harikrishna-patnala @nvazquez @weizhouapache

Comment (weizhouapache) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java:2825]: @shwstppr 
this uses same code as line 2810
we could change both lines if you agree this is better.

Comment (shwstppr) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java:2825]: @weizhouapache probably safe to use `firstHostDatastoreSystemMO.findDatastoreByName` then. I've not played with the managed storages myself.
From what I understand, HypervisorHostHelper.findDatastoreWithBackwardsCompatibility tries to find the datastore using custom field(cloud.uuid) set for the datastore. Not sure if we even do that for managed datastores.

**Review by shwstppr — APPROVED** [review]
code LGTM

File: plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java (hunk 2 above)
Comment (weizhouapache) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java:2825]: agree with what you said. @shwstppr 
too bad we do not have a managed storage to verify it

**Comment by shwstppr** [comment]
@blueorangutan test rocky8 vmware-67u3

**Comment by weizhouapache** [comment]
> @weizhouapache will we merge this? it is not harming anything as far as we can see, and only solidfire users can test it for us.

@DaanHoogland 
I am ok with merging this. but you know, I am the author :-)

**Comment by weizhouapache** [comment]
> > > @weizhouapache will we merge this? it is not harming anything as far as we can see, and only solidfire users can test it for us.
> > 
> > 
> > @DaanHoogland I am ok with merging this. but you know, I am the author :-)
> 
> yes, but you kept it in draft untill now, so I am checking with you anyway.

ah, moved it to "ready for review".
I will merge this if there is no objection until the end of Friday

**Comment by DaanHoogland** [comment]
> > > > @weizhouapache will we merge this? it is not harming anything as far as we can see, and only solidfire users can test it for us.
> > > 
> > > 
> > > @DaanHoogland I am ok with merging this. but you know, I am the author :-)
> > 
> > 
> > yes, but you kept it in draft untill now, so I am checking with you anyway.
> 
> ah, moved it to "ready for review". I will merge this if there is no objection until the end of Friday

ok, let's merge as you suggested but keep the issue open as @rohityadavcloud suggested until a solidfire user confirmed.

**Comment by rohityadavcloud** [comment]
@weizhouapache I've tested this fix backported to a 4.17 env with Solidfire and VMware; it fixes the issue of previously failured where it would attempt to create datastore on an already created datastore for a solidfire iSCSI target. However, it doesn't address the issue https://github.com/apache/cloudstack/issues/7148 completely.

**Pull request merged**